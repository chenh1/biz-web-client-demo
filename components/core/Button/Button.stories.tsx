import type { ButtonType } from './Button'
import { Button } from './Button';

export default {
  component: Button
}

type Args = {
  args: ButtonType
}

// More on writing stories with args: https://storybook.js.org/docs/react/writing-stories/args
export const Primary: Args = {
  args: {
    primary: true,
    label: 'Click me now',
  },
};

export const PrimaryAltColor1: Args = {
  args: {
    primary: true,
    color: "ultra-purple",
    label: 'Click me now',
  },
};

export const Secondary: Args = {
  args: {
    label: 'Click me now',
  },
};

export const Large: Args = {
  args: {
    size: 'lg',
    label: 'Click me now',
  },
};

export const Small: Args = {
  args: {
    size: 'sm',
    label: 'Click me now',
  },
};
