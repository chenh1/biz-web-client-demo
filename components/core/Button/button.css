.triangle { 
  position: absolute;
  width: 20px;
  top: 50%;
  display: none;
}

.secondary:hover .triangle {
  display: block;
}

.triangle.static {
  display: block;
}
/* 
@media (min-width: 768px) {
  .show-triangle .triangle {
    display: block;
  }

  .hide-triangle .triangle {
    display: none;
  }
} */

.triangle.left {
  left: 0;
  transform: translateX(-200%) rotate(90deg);
}

.triangle.right {
  right: 0;
  transform: translateX(200%) rotate(-90deg);
}

.triangle>div { 
  width: 20px;
  height: 2px;
  background: #050E1C;
}

.dark .triangle>div {
  background: #F3F4F6;
}

.triangle>div:before {
  content: " ";
  display: block;
  width: 20px;
  height: 2px;
  background: #050E1C;
  -webkit-transform: rotate(56deg);
  -moz-transform: rotate(56deg);
  -ms-transform: rotate(56deg);
  transform: rotate(56deg);
  position: absolute;
  top: -8px;
  right: -5px;
}
.dark .triangle>div:before {
  background: #F3F4F6;
}
.triangle>div:after {
  content: " ";
  display: block;
  width: 20px;
  height: 2px;
  background: #050E1C;
  -webkit-transform: rotate(-56deg);
  -moz-transform: rotate(-56deg);
  -ms-transform: rotate(-56deg);
  transform: rotate(-56deg);
  position: absolute;
  top: -8px;
  left: -5px;
}
.dark .triangle>div:after {
  background: #F3F4F6;
}