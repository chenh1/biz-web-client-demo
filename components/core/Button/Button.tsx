"use client"

import type { CoreColors } from '../../constants/colors'
import type { IncrementKeys } from '@/components/constants/types';

import React, { useState, useEffect, useMemo } from 'react';
import Link from 'next/link';
import { CLICK_CLASS, SECTION_VIEW_CLASS } from '../../constants/tagging';
import './button.css'

const getSizeClasses = (size: IncrementKeys): string => {
  switch (size) {
    case 'sm': {
      return 'px-4 py-2.5';
    }
    case 'lg': {
      return 'px-6 py-3';
    }
    default: {
      return 'px-5 py-2.5';
    }
  }
};

const getColorClasses = (color: CoreColors): string => {
  switch(color) {
    case 'ultra-purple': {
      return 'bg-ultra-purple border-ultra-purple dark:bg-ultra-purple dark:border-ultra-purple'
    }
    case 'hyper-red': {
      return 'bg-hyper-red border-hyper-red dark:bg-hyper-red dark:border-hyper-red'
    }
    default: {
      return "";
    }
  }
}

const getModeClasses = (isPrimary: boolean): string =>
  isPrimary
    ? 'font-headings relative text-max-white'
    : 'font-headings relative text-slate-700 bg-transparent border-slate-700 dark:text-max-white dark:border-max-white active:text-max-white active:bg-slate-700 active:border-slate-700 dark:active:bg-max-white dark:active:border-max-white dark:active:text-night-black';

const BASE_BUTTON_CLASSES: string =
  'w-full sm:w-auto cursor-pointer border-2 font-bold leading-none inline-block rounded-lg text-center';

export type ButtonType = { type?: string; disabled?: boolean; primary?: boolean, color?: CoreColors, size?: IncrementKeys, label: string, url?: string, textAlign?: string, onClick?: (any) => void, dataGtmId?: `gtm-${string}`, gtmClick?: `gtm-${string}`, gtmSectionView?: `gtm-${string}` }
/**
 * Primary UI component for user interaction
 */
export const Button: React.FC<ButtonType> = (
  { primary = false, color = "hyper-red", size = 'md', label, url, textAlign, onClick, dataGtmId, gtmClick, gtmSectionView, ...props }
) => {
  const [ isHover, setIsHover ] = useState(false);
  const computedClasses = useMemo(() => {
    const modeClass = getModeClasses(primary);
    const sizeClass = getSizeClasses(size);
    const textAlignClass = textAlign ? `text-${textAlign}` : '';
    return [modeClass, sizeClass, textAlignClass, url ? 'w-full' : ''].join(' ');
  }, [primary, size, textAlign, url]);

  const computedColorClasses = useMemo(() => {
    return primary ? getColorClasses(color) : '';
  }, [color])

  const ClickElement = !!url ?
    props => <Link data-testid="buttonLink" {...props}/> :
    props => <button data-testid="button" {...props}/>;

  return (
    <ClickElement 
      href={url}
      data-gtm-id={dataGtmId}
      type="button" 
      onMouseOver={() => !isHover && setIsHover(true)} 
      onMouseOut={() => setIsHover(false)} 
      onMouseLeave={() => setIsHover(false)} 
      className={`${gtmClick ? CLICK_CLASS : ''} ${!primary ? 'secondary' : ''} ${gtmSectionView ? SECTION_VIEW_CLASS : ''} ${BASE_BUTTON_CLASSES} ${computedClasses} ${computedColorClasses}`} 
      onClick={(e) => {
        onClick && onClick(e)
        //setIsHover(false);
      }}
      {...props}
    > 
      {primary && <div className={`${computedColorClasses} rounded-lg absolute left-0 top-0 w-full h-full ${isHover ? 'animate-plump' : ''}`}></div>}
      {!primary &&
        <>
          <div className="triangle left">
            <div></div>
          </div>
          <div className="triangle right">
            <div></div>
          </div>
        </>
      }
      <span className="pointer-events-none relative z-10">
        {label}
      </span>
    </ClickElement>
  );
};
