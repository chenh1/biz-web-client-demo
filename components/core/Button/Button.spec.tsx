import { Button } from "./Button";
import { fireEvent, render } from '@testing-library/react';
import { JSDOM } from "jsdom"
const dom = new JSDOM()
global.document = dom.window.document
global.window = dom.window


it('renders a button', async () => {
  const { getByTestId, asFragment } = render(
    <Button label="Click Me"/>
  );

  expect(asFragment()).toMatchSnapshot();
});


it('animates button on hover and reverts back after leave', async () => {
  const { getByTestId, asFragment } = render(
    <Button label="Click Me"/>
  );

  fireEvent.mouseEnter(getByTestId("button"))

  expect(asFragment()).toMatchSnapshot();

  fireEvent.mouseLeave(getByTestId("button"))

  expect(asFragment()).toMatchSnapshot();
});
