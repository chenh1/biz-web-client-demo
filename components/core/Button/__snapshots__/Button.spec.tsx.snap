// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`animates button on hover and reverts back after leave 1`] = `
<DocumentFragment>
  <button
    class=" secondary  w-full sm:w-auto cursor-pointer border-2 font-bold leading-none inline-block rounded-lg text-center font-headings relative text-slate-700 bg-transparent border-slate-700 dark:text-max-white dark:border-max-white active:text-max-white active:bg-slate-700 active:border-slate-700 dark:active:bg-max-white dark:active:border-max-white dark:active:text-night-black px-5 py-2.5   "
    data-testid="button"
    type="button"
  >
    <div
      class="triangle left"
    >
      <div />
    </div>
    <div
      class="triangle right"
    >
      <div />
    </div>
    <span
      class="pointer-events-none relative z-10"
    >
      Click Me
    </span>
  </button>
</DocumentFragment>
`;

exports[`animates button on hover and reverts back after leave 2`] = `
<DocumentFragment>
  <button
    class=" secondary  w-full sm:w-auto cursor-pointer border-2 font-bold leading-none inline-block rounded-lg text-center font-headings relative text-slate-700 bg-transparent border-slate-700 dark:text-max-white dark:border-max-white active:text-max-white active:bg-slate-700 active:border-slate-700 dark:active:bg-max-white dark:active:border-max-white dark:active:text-night-black px-5 py-2.5   "
    data-testid="button"
    type="button"
  >
    <div
      class="triangle left"
    >
      <div />
    </div>
    <div
      class="triangle right"
    >
      <div />
    </div>
    <span
      class="pointer-events-none relative z-10"
    >
      Click Me
    </span>
  </button>
</DocumentFragment>
`;

exports[`renders a button 1`] = `
<DocumentFragment>
  <button
    class=" secondary  w-full sm:w-auto cursor-pointer border-2 font-bold leading-none inline-block rounded-lg text-center font-headings relative text-slate-700 bg-transparent border-slate-700 dark:text-max-white dark:border-max-white active:text-max-white active:bg-slate-700 active:border-slate-700 dark:active:bg-max-white dark:active:border-max-white dark:active:text-night-black px-5 py-2.5   "
    data-testid="button"
    type="button"
  >
    <div
      class="triangle left"
    >
      <div />
    </div>
    <div
      class="triangle right"
    >
      <div />
    </div>
    <span
      class="pointer-events-none relative z-10"
    >
      Click Me
    </span>
  </button>
</DocumentFragment>
`;
